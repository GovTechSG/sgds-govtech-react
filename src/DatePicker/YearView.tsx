import classNames from 'classnames';
import * as React from 'react';

export interface YearViewProps extends React.HTMLAttributes<HTMLElement> {
  displayDate: Date;
  onClickYear: Function;
}

export const YearView = React.forwardRef<HTMLDivElement, YearViewProps>(
  ({ displayDate, onClickYear, ...props }, ref) => {
    const displayYear = displayDate.getFullYear();
    const startLimit = displayYear - 5;
    const endLimit = displayYear + 6;
    const yearArray = [];

  for (let i = startLimit; i < endLimit + 1; i++) {
      yearArray.push(i);
    } 

    return (
      <div className="container text-center" ref={ref} {...props}>
        <div className="row">
          {yearArray.map((y) => (
            <div
              className={classNames(displayYear === y && 'active', 'col-4')}
              key={y}
            >
              <button onClick={() => onClickYear(y)}>{y}</button>
            </div>
          ))}
        </div>
      </div>
    );
  }
);

export default YearView;
