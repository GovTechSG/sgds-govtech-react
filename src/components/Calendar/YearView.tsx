import classNames from 'classnames';
import * as React from 'react';
import { useState } from 'react';
import Button from '../Button/Button';
import Col from '../Col';
export interface YearViewProps extends React.HTMLAttributes<HTMLElement> {
  displayDate: Date;
}



export const YearView = React.forwardRef<HTMLDivElement, YearViewProps>(
  ({ displayDate, ...props }, ref) => {
      console.log(displayDate.getFullYear())
      const displayYear = displayDate.getFullYear()
      const startLimit = displayYear - 5
      const endLimit = displayYear + 6
      const yearArray = []

      for(let i = startLimit; i<endLimit + 1; i++){
          yearArray.push(i)
      }
    // const [year, setYear] = useState(displayDate.getFullYear())
    // const [startLimit, setStartLimit]  = useState(displayDate.getFullYear() - 5)
    // const [endLimit, setEndLimit] = useState(displayDate.getFullYear() + 6)
    return (
      <div className="container text-center" ref={ref} {...props}>
        <div className="row">
          {yearArray.map((y) => (
            <div
              className={classNames(displayYear === y && 'active', 'col-4')}
              key={y}
            >
              <button>{y}</button>
            </div>
          ))}
        </div>
      </div>
    );
  }
);

export default YearView;
