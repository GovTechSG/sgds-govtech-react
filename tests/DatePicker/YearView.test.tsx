import * as React from 'react';
import { fireEvent, render } from '@testing-library/react';
import YearView from '../../src/DatePicker/YearView'

describe('YearView', () => {
    const mockFn = jest.fn()
    const displayDate = new Date()
    it ('should have default html structure', ()=> {
      
        const {asFragment, container} = render(<YearView onClickYear={mockFn} displayDate={displayDate} />)

        expect(asFragment()).toMatchSnapshot()
        expect(container.querySelector('div.container.text-center')).toBeInTheDocument()
        expect(container.querySelectorAll('div.row').length).toEqual(1)
        expect(container.querySelectorAll('div.col-4').length).toEqual(12)
        expect(container.querySelectorAll('button').length).toEqual(12)
        expect(container.querySelectorAll('div.active').length).toEqual(1) 
        expect(container.querySelector('div.active')?.textContent).toEqual(`${displayDate.getFullYear()}`) 

    }) 

    it('given current year, min Year displayed is current-5 and max year is current+5', () => {
        const currentYear = displayDate.getFullYear()
        const {getByText, queryByText} = render(<YearView onClickYear={mockFn} displayDate={displayDate} />)

        expect(getByText(`${currentYear}`)).toBeInTheDocument()
        expect(getByText(`${currentYear-5}`)).toBeInTheDocument()
        expect(queryByText(`${currentYear-6}`)).not.toBeInTheDocument()
        expect(getByText(`${currentYear+6}`)).toBeInTheDocument()
        expect(queryByText(`${currentYear+7}`)).not.toBeInTheDocument() 
    })

    it('onClick handler should work', () => {
        const { container} = render(<YearView onClickYear={mockFn} displayDate={displayDate} />)
        expect(container.querySelectorAll('button').length).toEqual(12)

        const buttonOne = container.querySelectorAll('button')[0]

        fireEvent.click(buttonOne)

        expect(mockFn).toHaveBeenCalled() 
    }) 
})