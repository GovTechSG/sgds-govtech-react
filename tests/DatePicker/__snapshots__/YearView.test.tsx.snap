// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YearView should have default html structure 1`] = `
<DocumentFragment>
  <div
    class="container text-center"
  >
    <div
      class="row"
    >
      <div
        class="col-4"
      >
        <button>
          2017
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          2018
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          2019
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          2020
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          2021
        </button>
      </div>
      <div
        class="active col-4"
      >
        <button>
          2022
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          2023
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          2024
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          2025
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          2026
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          2027
        </button>
      </div>
      <div
        class="col-4"
      >
        <button>
          2028
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
